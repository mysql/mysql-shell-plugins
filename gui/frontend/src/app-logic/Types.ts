/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

// TODO: check if we can switch to camel case for those DB data types that use snake case.

/* eslint-disable @typescript-eslint/naming-convention */

import * as codicon from "../components/ui/Codicon";

// Types used in different places in the application (modules, parser, scripting etc.).

export interface IDictionary {
    [key: string]: unknown;
}

// Types of columns in a table column definition, which we support in the application.
// Also used for column affinities in SQLite (instead of column types).
export enum DBDataType {
    Unknown,

    // Integer types.
    TinyInt,
    SmallInt,
    MediumInt,
    Int,                // Any integer like number type (numeric is essentially the same) in SQLite.
    Bigint,
    UInteger,           // Mapped type from the shell.

    // Float types.
    Float,
    Real,               // Any floating point type in SQLite.
    Double,
    Decimal,

    // Binary types.
    Binary,
    Varbinary,

    // Text types.
    Char,
    Nchar,
    Varchar,
    Nvarchar,

    String,             // Mapped type from the shell.
    TinyText,
    Text,               // Any text in SQLite.
    MediumText,
    LongText,

    // Blob types.
    TinyBlob,
    Blob,               // NULL or no (explicit) type in SQLite.
    MediumBlob,
    LongBlob,

    // Date and time types.
    DateTime,
    DateTime_f,
    Date,
    Time,
    Time_f,
    Year,
    Timestamp,
    Timestamp_f,

    // Geometry types.
    Geometry,
    Point,
    LineString,
    Polygon,
    GeometryCollection,
    MultiPoint,
    MultiLineString,
    MultiPolygon,

    // Special types.
    Numeric,            // SQLite only, similar to INTEGER.
    Json,
    Bit,
    Boolean,
    Enum,
    Set,
}

// Describes the number of parameters for a DB data type. It doesn't say anything about the parameter types, though.
export enum ParameterFormatType {
    None = 0,
    One = 1,            // (n)
    OneOrZero = 2,      // [(n)]
    Two = 3,            // (m,n)
    TwoOrOne = 4,       // (m[,n])
    TwoOrZero = 5,      // [(m,n)]
    TwoOrOneOrZero = 6, // [(m[, n])]
    List = 10,          // (m, n, o, p ...)
}

export interface IDBCharacterSet {
    collations: string[];
    defaultCollation: string;
    description: string;
}

// Describes a data type in a database.
export interface IDBDataTypeDetails {
    type: DBDataType;

    characterMaximumLength?: number;
    characterOctetLength?: number;
    dateTimePrecision?: number;
    flags?: string[];
    numericPrecision?: number;
    numericPrecisionRadix?: number;
    numericScale?: number;
    needsQuotes?: boolean;
    parameterFormatType?: ParameterFormatType;
    synonyms?: string[];
}

export interface IColumnInfo {
    /** The column's title shown in the UI. */
    title: string;

    /** The name of the field in the data, which contains the data for this column. */
    field: string;

    /** Note: SQLite has no column data types like other RDBMS-es. Instead we store column affinities here. */
    dataType: IDBDataTypeDetails;
}

export enum MessageType {
    Error,
    Warning,
    Info,
    Interactive,
    Response,
}

export interface IExecutionInfo {
    type?: MessageType;
    text: string;
}

// Types for requesting a specific dialog and sending back the entered values.

export enum DialogType {
    Prompt,     // A simple prompt value dialog, requesting a single value from the user.
    Confirm,    // Confirm a question (yes, no, alt).
    Select,     // Select one entry from a list.
    MrsService, // A dialog to create or edit an MRS service.
    MrsSchema,
}

/**
 * A set of values that describe a single modal dialog request.
 */
export interface IDialogRequest extends IDictionary {
    type: DialogType;

    /** An id to identify the invocation. */
    id: string;

    /** Optionally used to set a customized dialog title (where supported). */
    title?: string;

    /** A list of strings to be rendered before the actual prompt, each in an own paragraph. */
    description?: string[];

    /** Values to configure how the dialog looks like (available options in drop downs etc.). */
    parameters?: IDictionary;

    /** Values to pre-fill certain elements or additional data for captions, payload etc. */
    values?: IDictionary;

    /** Additional data for specific use cases. */
    data?: IDictionary;
}

/**
 * What decision made the user to close a dialog.
 */
export enum DialogResponseClosure {
    Accept,
    Decline,
    Alternative,

    /** Set when no decision was made. */
    Cancel,
}

export interface IDialogResponse extends IDictionary {
    id: string;
    type: DialogType;
    closure: DialogResponseClosure;

    values?: IDictionary;
    data?: IDictionary;
}

export interface IStatusbarInfo {
    id: string;
    text?: string;
    icon?: string | codicon.Codicon;
    choices?: Array<{ label: string; data: IDictionary }>;
    visible?: boolean;
    standout?: boolean;
}

/**
 * Defines the structure for password requests sent by various access methods. Passwords are stored by the backend
 * using native OS methods.
 */
export interface IServicePasswordRequest {
    /** The id of the request that asked for the password. */
    requestId: string;

    /** The password dialog title. */
    caption?: string;

    /** A list of strings describing the circumstances of the password request. */
    description?: string[];

    /** A human readable string describing the service. */
    service?: string;

    /** The user name for which to get the password. */
    user?: string;

    /** Any other data the caller wants to pass over. */
    payload?: IDictionary;
}
