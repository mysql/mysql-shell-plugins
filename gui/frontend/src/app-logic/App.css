/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

@viewport {
    width: device-width;
    zoom: 1.0;
}

@font-face {
    font-family: "SourceCodePro+Powerline+Awesome+MySQL Regular";
    font-style: normal;
    font-weight: 400;
    src: url(../assets/fonts/MySQL.ttf) format("truetype");
}

@font-face {
    font-family: "SourceCodePro+Powerline+Awesome+MySQL";
    font-style: normal;
    font-weight: 400;
    src: url(../assets/fonts/MySQL.ttf) format("truetype");
}

* {
    box-sizing: border-box;
}

html,
body {
    min-height: 100vh;
    margin: 0;
}

body {
    display: flex;
    flex-direction: column;
    margin: 0;
    padding: 0;
    overflow: hidden;

    color: var(--foreground);
    background-color: transparent;

    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
    font-weight: 400;
    font-size: 15px;

    user-select: none;
}

#root {
    display: flex;
    flex-direction: column;
    height: 100vh;

    background-color: transparent;
}

a {
    color: var(--textLink-foreground);
}

a:active {
    color: var(--textLink-activeForeground);
}

pre {
    color: var(--textPreformat-foreground);
}

.verticalCenterContent {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.centerContent {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
}

.ellipsis {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.msg .dropShadow {
    filter: drop-shadow(3px 3px 5px var(--widget-shadow));
}

.copyright::after {
    content: "Â© 2020, 2022, Oracle Corporation and/or its affiliates.";
    display: inline-block;
    margin-top: 80px;
    font-size: 10px;
    font-weight: 100;
    color: var(--text-dimmedForeground);
}

html.themeSwitch,
html.themeSwitch *,
html.themeSwitch *:before,
html.themeSwitch *:after {
    transition: color 500ms, background-color 500ms !important;
    transition-delay: 0 !important;
}

.msg .manualFocus,
.msg [tabindex="0"]:focus,
.msg input[type=button]:focus,
.msg input[type=text]:focus,
.msg input[type=password]:focus,
.msg input[type=number]:focus,

.msg button[search-type=option]:focus,
.msg.searchContainer:focus-within,

.msg select:focus,
.msg textarea:focus,
.msg button:focus {
    outline-width: 1px;
    outline-style: solid;
    outline-offset: -1px;
    outline-color: var(--focusBorder);
    opacity: 1 !important
}

/* The next two rules disable the number input spin buttons for all browsers. */
.msg input[type="number"] {
    -webkit-appearance: textfield;
    -moz-appearance: textfield;
    appearance: textfield;
}

.msg input[type=number]::-webkit-inner-spin-button,
.msg input[type=number]::-webkit-outer-spin-button {
    -webkit-appearance: none;
}

.msg::selection {
    background-color: var(--selection-background);
}

.msg.applicationHost {
    overflow: hidden;
    align-items: stretch;
    height: 100vh;

    background-color: var(--background);
}

.msg.applicationHost .moduleHost {
    display: none;
}

.msg.applicationHost .moduleHost.active {
    display: flex;
}

/* Special styles for macOS auto disappearing scroll bars. */
.msg .fixedScrollbar {
    overflow: auto;

    -webkit-overflow-scrolling: touch;

    scrollbar-width: auto;
    scrollbar-color: var(--scrollbarSlider-background) var(--scrollbarSlider-shadow);
}

.msg .fixedScrollbar::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 7px;
    height: 7px;

    background-color: var(--scrollbarSlider-shadow);
}

.msg .fixedScrollbar::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: var(--scrollbarSlider-background);
}

.msg .fixedScrollbar::-webkit-scrollbar-corner {
    background-color: var(--scrollbarSlider-shadow);
}

/* Some browser specific rules. */

/* Removes the additional dotted focus indicator only visible in Firefox. */
::-moz-focus-inner {
    border: 0;
}
