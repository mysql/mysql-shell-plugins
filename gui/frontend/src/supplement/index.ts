/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mysqlInfo, sqliteInfo } from "../app-logic/RdbmsInfo";
import { DBDataType, IColumnInfo, IDictionary } from "../app-logic/Types";
import { DBType } from "./ShellInterface";

// Commonly used data types and functions.

/**
 * These are the supported languages in the code editor.
 * This includes our mixed language (msg), which combines SQL, Python, JS and TS in one editor.
 */
export type EditorLanguage = (
    "text" | "typescript" | "javascript" | "mysql" | "sql" | "python" | "json" | "markdown" | "msg" | "xml" | "ini"
);

export interface ITextRange {
    readonly startLine: number;
    readonly startColumn: number;
    readonly endLine: number;
    readonly endColumn: number;
}

export interface IExecutionContext {
    id: string;
    code: string;
    codeLength: number;
    language: EditorLanguage;
    isSQLLike: boolean;
    linkId?: number;
}

export interface ISqlPageRequest {
    context: IExecutionContext;
    oldRequestId: string;
    page: number;
    sql: string;
}

export interface IRunQueryRequest {
    data: IDictionary;
    query: string;
    parameters: Array<[string, string]>;
    linkId: number;
}

export interface IScriptRequest {
    /** A unique ID to identify the script in this request. */
    scriptId: string;

    /** The language of the script. */
    language: EditorLanguage;

    name?: string;
    content: string;
}

/** Used to request the creation of a new script in the extension. */
export interface INewScriptRequest {
    /** The id of the page (connection tab) to open the new script in. */
    page: string;

    /** The language to use for the new script file. */
    language: EditorLanguage;

    /** Optional content to use for the new script file. */
    content?: string;
}

export { Stack } from "./Stack";

export type WorkerExecutorType<T> = (
    onResult?: ((taskId: number, value: T) => void),
    onError?: ((taskId: number, reason?: unknown) => void)
) => void;

// A callback with a promise-like callback function.
export interface IThenableCallback<T> {
    then: (onResult?: ((taskId: number, value: T) => void)) => IThenableCallback<T>;
    catch: (onError?: ((taskId: number, reason?: unknown) => void)) => void;
}

// This conditional type enforces type widening on generic parameter types.
export type ValueType<T> = T extends string
    ? string
    : T extends number
        ? number
        : T extends boolean
            ? boolean
            : T extends undefined
                ? undefined
                : [T] extends [unknown]
                    ? T
                    : object;

/**
 * Generates a generic column info record for each raw column returned from a request. The format of the
 * raw columns depends on the used RDBMS.
 *
 * @param dbType Determines for which DB type the conversion happens (MySQL, SQLite and so on).
 * @param rawColumns Column info as returned by the backend.
 * @param useName Use column name not index.
 *
 * @returns A list of columns with RDBMS-agnostic details.
 */
export const generateColumnInfo = (dbType: DBType, rawColumns?: Array<{ name: string; type: string; length: number }>,
    useName?: boolean): IColumnInfo[] => {
    if (!rawColumns) {
        return [];
    }

    const dataTypes = dbType === DBType.MySQL ? mysqlInfo.dataTypes : sqliteInfo.dataTypes;

    return rawColumns.map((entry, index) => {
        let type;
        if (entry.type.toLowerCase() === "bytes") {
            // For now, use length to switch between binary and blob
            if (entry.length < 256) {
                type = DBDataType.Binary;
            } else {
                type = DBDataType.Blob;
            }
        } else {
            const foundType = dataTypes.get(entry.type.toLowerCase());
            type = foundType ? foundType.type : DBDataType.Unknown;
        }

        return {
            title: entry.name,
            field: useName ? entry.name : String(index),
            dataType: {
                type,
            },
        };
    });
};

/**
 * Converts the given rows into a dictionary format as required by the UI, if it isn't already in that format.
 *
 * @param columns Column informations to find column IDs.
 * @param rows The rows to convert.
 *
 * @returns The converted rows.
 */
export const convertRows = (columns: IColumnInfo[], rows?: unknown[]): IDictionary[] => {
    if (!rows || rows.length === 0) {
        return [];
    }

    if (Array.isArray(rows[0])) {
        // Data is array of arrays of fields.
        const convertedRows: IDictionary[] = [];
        rows.forEach((value): void => {
            const row: IDictionary = {};

            const entry = value as unknown[];
            columns.forEach((column: IColumnInfo, columnIndex: number): void => {
                row[column.field] = entry[columnIndex];
            });
            convertedRows.push(row);
        });

        return convertedRows;
    }

    return rows as IDictionary[];
};

