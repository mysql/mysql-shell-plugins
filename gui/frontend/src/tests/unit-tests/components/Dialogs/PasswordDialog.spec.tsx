/*
 * Copyright (c) 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";
import { mount } from "enzyme";
import keyboardKey from "keyboard-key";

import { PasswordDialog } from "../../../../components/Dialogs";
import { changeInputValue, nextProcessTick, sendKeyPress, snapshotFromWrapper } from "../../test-helpers";
import { requisitions } from "../../../../supplement/Requisitions";
import { IServicePasswordRequest } from "../../../../app-logic/Types";

describe("Password Dialog Tests", (): void => {
    it("Render Test", () => {
        const component = mount<PasswordDialog>(
            <PasswordDialog />,
        );

        expect(snapshotFromWrapper(component)).toMatchSnapshot();
        component.unmount();
    });

    it("Show and Cancel Dialog", async () => {
        const component = mount<PasswordDialog>(
            <PasswordDialog />,
        );

        const acceptPassword = jest.fn(
            (_result: { request: IServicePasswordRequest; password: string }): Promise<boolean> => {
                return Promise.resolve(true);
            },
        );

        const cancelPassword = jest.fn((_request: IServicePasswordRequest): Promise<boolean> => {
            return Promise.resolve(true);
        });

        requisitions.register("acceptPassword", acceptPassword);
        requisitions.register("cancelPassword", cancelPassword);

        let portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);

        const request: IServicePasswordRequest = {
            requestId: "123",
            caption: "Enter Password",
            description: ["I need your password"],
            service: "unit tests",
            user: "mike",
            payload: { extra: 123 },
        };
        await requisitions.execute("requestPassword", request);
        await nextProcessTick();

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(1);
        expect(portals[0]).toMatchSnapshot();

        sendKeyPress(keyboardKey.Escape);
        await nextProcessTick();

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);
        expect(acceptPassword).toHaveBeenCalledTimes(0);
        expect(cancelPassword).toHaveBeenCalledTimes(1);
        expect(cancelPassword).toHaveBeenCalledWith(request);

        component.unmount();
    });

    it("Show and Accept Dialog", async () => {
        const component = mount<PasswordDialog>(
            <PasswordDialog />,
        );

        const acceptPassword = jest.fn(
            (_result: { request: IServicePasswordRequest; password: string }): Promise<boolean> => {
                return Promise.resolve(true);
            },
        );

        const cancelPassword = jest.fn((_request: IServicePasswordRequest): Promise<boolean> => {
            return Promise.resolve(true);
        });

        requisitions.register("acceptPassword", acceptPassword);
        requisitions.register("cancelPassword", cancelPassword);

        let portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);

        const request: IServicePasswordRequest = {
            requestId: "123",
            description: ["I need your password"],
            service: "unit tests",
            user: "mike",
            payload: { extra: 123 },
        };
        await requisitions.execute("requestPassword", request);
        await nextProcessTick();

        expect(component.state().password).toBe("");

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(1);
        expect(portals[0]).toMatchSnapshot();

        const inputs = portals[0].getElementsByTagName("input");
        expect(inputs).toHaveLength(1);
        changeInputValue(inputs[0], "swordfish");
        await nextProcessTick();
        expect(component.state().password).toBe("swordfish");

        // Close dialog using the enter key in the password field.
        sendKeyPress(keyboardKey.Enter, inputs[0]);
        await nextProcessTick();

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);
        expect(cancelPassword).toHaveBeenCalledTimes(0);
        expect(acceptPassword).toHaveBeenCalledTimes(1);

        expect(acceptPassword).toHaveBeenCalledWith({ request, password: "swordfish" });

        // Open the dialog again and close it using the OK button.
        await requisitions.execute("requestPassword", request);
        await nextProcessTick();

        const buttons = portals[0].getElementsByTagName("button");
        expect(buttons).toHaveLength(3);
        expect(buttons[1].id).toBe("ok");
        expect(buttons[2].id).toBe("cancel");
        buttons[1].click();
        await nextProcessTick();

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);
        expect(cancelPassword).toHaveBeenCalledTimes(0);
        expect(acceptPassword).toHaveBeenCalledTimes(2);

        // No password was set, so no password is returned.
        expect(acceptPassword).toHaveBeenLastCalledWith({ request, password: "" });

        component.unmount();
    });
});
