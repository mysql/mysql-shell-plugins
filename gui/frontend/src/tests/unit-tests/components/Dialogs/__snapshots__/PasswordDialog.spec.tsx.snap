// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Dialog Tests Render Test 1`] = `<PasswordDialog />`;

exports[`Password Dialog Tests Show and Accept Dialog 1`] = `
<div
  __self="[object Object]"
  __source="[object Object]"
  class="msg portal"
>
  <div
    __self="[object Object]"
    __source="[object Object]"
    class="msg dialog visible passwordDialog"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="title verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg icon codicon codicon-lock"
      />
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip="expand"
      >
        Enter Password
      </label>
      <button
        __self="[object Object]"
        __source="[object Object]"
        class="msg button imageOnly"
        id="closeButton"
        tabindex="0"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg icon"
          style="mask-image: url(close2.svg);"
        />
      </button>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="content fixedScrollbar"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg grid"
        style="grid-template-columns: 30% auto; row-gap: 8px; column-gap: 8px;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; align-items: flex-start; grid-column: span 2;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultText"
            id="caption"
            language="ansi"
            style="flex-direction: column;"
          >
            <code>
              <span>
                I need your password
              </span>
            </code>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell left"
          style="flex-direction: row; align-items: center;"
        >
          Service:
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell right"
          style="flex-direction: row; align-items: center;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultText"
            language="ansi"
            style="flex-direction: column;"
          >
            <code>
              <span>
                unit tests
              </span>
            </code>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell left"
          style="flex-direction: row; align-items: center;"
        >
          User Name:
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell right"
          style="flex-direction: row; align-items: center;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultText"
            language="ansi"
            style="flex-direction: column;"
          >
            <code>
              <span>
                mike
              </span>
            </code>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell left"
          style="flex-direction: row; align-items: center;"
        >
          Password:
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell right"
          style="flex-direction: row; align-items: center;"
        >
          <input
            __self="[object Object]"
            __source="[object Object]"
            class="msg input"
            spellcheck="true"
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="footer verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="OK"
          class="msg button"
          id="ok"
          tabindex="0"
        >
          OK
        </button>
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="Cancel"
          class="msg button"
          id="cancel"
          tabindex="0"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Password Dialog Tests Show and Cancel Dialog 1`] = `
<div
  __self="[object Object]"
  __source="[object Object]"
  class="msg portal"
>
  <div
    __self="[object Object]"
    __source="[object Object]"
    class="msg dialog visible passwordDialog"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="title verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg icon codicon codicon-lock"
      />
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip="expand"
      >
        Enter Password
      </label>
      <button
        __self="[object Object]"
        __source="[object Object]"
        class="msg button imageOnly"
        id="closeButton"
        tabindex="0"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg icon"
          style="mask-image: url(close2.svg);"
        />
      </button>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="content fixedScrollbar"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg grid"
        style="grid-template-columns: 30% auto; row-gap: 8px; column-gap: 8px;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; align-items: flex-start; grid-column: span 2;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultText"
            id="caption"
            language="ansi"
            style="flex-direction: column;"
          >
            <code>
              <span>
                I need your password
              </span>
            </code>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell left"
          style="flex-direction: row; align-items: center;"
        >
          Service:
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell right"
          style="flex-direction: row; align-items: center;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultText"
            language="ansi"
            style="flex-direction: column;"
          >
            <code>
              <span>
                unit tests
              </span>
            </code>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell left"
          style="flex-direction: row; align-items: center;"
        >
          User Name:
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell right"
          style="flex-direction: row; align-items: center;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultText"
            language="ansi"
            style="flex-direction: column;"
          >
            <code>
              <span>
                mike
              </span>
            </code>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell left"
          style="flex-direction: row; align-items: center;"
        >
          Password:
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell right"
          style="flex-direction: row; align-items: center;"
        >
          <input
            __self="[object Object]"
            __source="[object Object]"
            class="msg input"
            spellcheck="true"
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="footer verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="OK"
          class="msg button"
          id="ok"
          tabindex="0"
        >
          OK
        </button>
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="Cancel"
          class="msg button"
          id="cancel"
          tabindex="0"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
